File: src/main/java/com/fulcrumgenomics/personal/nhomer/GenerateRegionsFromFasta.java
/**
 * Copyright (c) 2016, Fulcrum Genomics LLC
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */


package com.fulcrumgenomics.personal.nhomer;

import com.fulcrumgenomics.cmdline.Personal;
import htsjdk.samtools.SAMSequenceDictionary;
import htsjdk.samtools.SAMSequenceRecord;
import htsjdk.samtools.reference.ReferenceSequenceFile;
import htsjdk.samtools.reference.ReferenceSequenceFileFactory;
import htsjdk.samtools.util.IOUtil;
import picard.cmdline.CommandLineProgram;
import picard.cmdline.CommandLineProgramProperties;
import picard.cmdline.Option;
import picard.cmdline.StandardOptionDefinitions;

import java.io.File;

@CommandLineProgramProperties(
        usage = "Generates a list of freebayes/bamtools region specifiers on stdout.",
        usageShort = "Generates a list of freebayes/bamtools region specifiers on stdout.",
        programGroup = Personal.class
)
public class GenerateRegionsFromFasta extends CommandLineProgram {

    @Option(shortName = StandardOptionDefinitions.INPUT_SHORT_NAME, doc = "Input FASTA.")
    public File INPUT;

    @Option(shortName = "W", doc = "The size of regions to output.")
    public int REGION_SIZE = 100000;

    @Override
    public int doWork() {
        IOUtil.assertFileIsReadable(INPUT);

        final ReferenceSequenceFile referenceSequenceFile = ReferenceSequenceFileFactory.getReferenceSequenceFile(INPUT);
        final SAMSequenceDictionary sequenceDictionary = referenceSequenceFile.getSequenceDictionary();

        for (int i = 0; i < sequenceDictionary.size(); i++) {
            final SAMSequenceRecord sequenceRecord = sequenceDictionary.getSequence(i);
            int regionStart = 0; // yes, this must be zero-based
            while (regionStart < sequenceRecord.getSequenceLength()) {
                final int regionEnd = Math.min(regionStart + REGION_SIZE, sequenceRecord.getSequenceLength());
                System.out.println(sequenceRecord.getSequenceName() + ":" + regionStart + "-" + regionEnd);
                regionStart = regionEnd;
            }
        }

        return 0;
    }
}

File: src/main/scala/com/fulcrumgenomics/personal/nhomer/GenerateRegionsFromFasta.scala
/*
 * The MIT License
 *
 * Copyright (c) 2016 Fulcrum Genomics LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.fulcrumgenomics.personal.nhomer

import java.nio.file.Path

import com.fulcrumgenomics.cmdline.{JeanLucTool, Personal}
import dagr.commons.CommonsDef.PathToFasta
import dagr.commons.io.Io
import dagr.commons.util.LazyLogging
import dagr.sopt._
import htsjdk.samtools.reference.ReferenceSequenceFileFactory

import scala.collection.JavaConversions._

@clp(
  description="Generates a list of freebayes/bamtools region specifiers.",
  group = classOf[Personal]
)
class GenerateRegionsFromFasta
( @arg(doc = "The input FASTA.") val input: PathToFasta,
  @arg(doc = "The output.") val output: Path = Io.StdOut,
  @arg(doc = "The size of the regions to output.") val regionSize: Int = 100000
) extends JeanLucTool with LazyLogging {

  Io.assertReadable(input)
  Io.assertCanWriteFile(output)

  override def execute: Int = {
    val referenceSequenceFile = ReferenceSequenceFileFactory.getReferenceSequenceFile(input.toFile)
    val sequenceDictionary = referenceSequenceFile.getSequenceDictionary
    sequenceDictionary.getSequences.foreach { sequenceRecord =>
      for (regionStart <- 0 until sequenceRecord.getSequenceLength by regionSize) {
        val regionEnd = Math.min(regionStart + regionSize, sequenceRecord.getSequenceLength)
       println(s"${sequenceRecord.getSequenceName}:$regionStart-$regionEnd")
      }
    }
    0
  }
}
